### ISSUE TITLE
[DOCS ISSUE] Better document MFS and it's relation with non-MFS operations

### ISSUE THREAD

**Issue by hsanjuan** [ipfs/docs#481]
Tracking a user request at https://github.com/ipfs/go-ipfs/issues/7084#issuecomment-608552812

**URL of the page in question:**

Maybe: https://docs.ipfs.io/guides/concepts/mfs/

**What's wrong with this page?**

See confusion in above thread.

Related: how to manually create and modify unixfs directories (it has come up several times in the last week).

**Comment by tmbdev** [comment]
The doc says:

> Because files in IPFS are content-addressed and immutable, they can be complicated to edit. Mutable File System (MFS) is a tool built into IPFS that lets you treat files like you would a normal name-based filesystem — you can add, remove, move, and edit MFS files and have all the work of updating links and hashes taken care of for you.

So, apparently, I can run things like:

```
$ ipfs files mkdir /foo
$ ipfs files ls -l /foo
$ echo hello | ipfs files write --create /foo/world
$ ipfs files ls -l /foo # hash has changed
$ ipfs files read /foo/world
```

That's nice. Somewhere, there is a mapping from path names to hashes.  The documentation says nothing about how this mapping is established or where it is stored.

My first thought was that `ipfs files` maintains a local key/value store that eventually gets turned into an mDAG. But apparently `ipfs files` refers to a rooted tree stored under the hash one gets from `ipfs files stat /`. Apparently, this root node can be changed with `ipfs files chcid`. None of that is obvious from the (minimal) documentation.

My suggestion would be to update the documentation to something like this:

"The `ipfs files` commands allows a Merkle DAG rooted at a given CID to be manipulated almost as if it were a file system. For that, subcommands of `ipfs files` mirror common UNIX commands. While traditional UNIX commands modify stateful data structures, their `ipfs files` equivalents simply return a new root CID that reflects the altered file system state; the original state is still available under the original CID. The `ipfs files` commands are a simple set of convenience functions that keep track of file system state by keeping track of the root CID; but the Merkle DAG the generate under to root CID is identical to one created with the basic commands."

I'd also suggest rethinking the `ipfs files` commands, because the way they work adds to the confusion. First of all, instead of "ipfs files chcid", I'd suggest using an environment variable "IPFS_FILES_ROOT=..." to define a file that contains the current root to be used by all `ipfs files` commands; that file is updated by those commands. The reason is that it is unclear how "ipfs files chcid" works when used from multiple different processes at the same time.

Also, while I understand the attraction of a single, unified namespace for IPFS, I think you need some way of integrating the UNIX and IPFS file system name spaces, otherwise, you end up with a profusion of commands depending on which paths are IPFS paths and which paths are UNIX paths.

You could address this by making everything under "/" IPFS by default, but reserving "/fs/..." for local files. Alternatively, you could make everything under "/" the local file system by default, but reserve "/ipfs/..." for IPFS paths. I also think it's a good idea for all commands to understand URL schemes (file:, ipfs:, etc.) for completely unambiguous specifications. That way, "ipfs files cp" could replace read, write, and cp. In addition, "ipfs files cp" and other commands like it should probably understand "-" as a shorthand for stdin/stdout.

**Comment by hsanjuan** [comment]
Two quick notes (agreeing on most otherwise):

> Apparently, this root node can be changed with `ipfs files chcid`.

According to the cli docs, this only changes the CID version or hash function of the root node of a given path.

> The documentation says nothing about how this mapping is established or where it is stored.

I would be against providing imlementation details in user documentation. How that happens does not (or should not) affect the usage of the feature.

**Comment by tmbdev** [comment]
> would be against providing imlementation details in user documentation. How that happens does not (or should not) affect the usage of the feature.

For using the feature, it's important to know where/how the root is stored and updated , since users may need to back it up, version control it, reason about concurrent access, etc. That is, you may decide that the storage format itself is an implementation detail, but the fact that it is (or is not) stored in the file system somewhere is something users have to deal with and that you can't hide.

**Comment by hsanjuan** [comment]
There is no difference between that or any ipfs related data, everything is in `~/.ipfs`.

>  (i.e., after a local file system restore, a directory on IPFS created with these tools would seem to revert as well, an unexpected behavior for a distributed file system).

Did you get the idea that somehow MFS is backed up for you somewhere else so that it would keep state even if you reverted your IPFS repo?

**Comment by tmbdev** [comment]
> There is no difference between that or any ipfs related data, everything is in ~/.ipfs.

MFS could store state in memory/a daemon, in the current working directory, or in IPNS, all places IPFS also already stores state. All of those could be reasonable choices for something like MFS, depending on which use cases you have in mind.

**Comment by hsanjuan** [comment]
@tmbdev thanks, I see how being more detailed there can help users.

@johnnymatthews I think we have enough feedback to actually write a good guide on MFS.

Do you think this should still be under concepts/MFS, or a different content location? I volunteer myself to write it.

**Comment by johnnymatthews** [comment]
Yeah `/concepts/mfs` works well for the URL. The title of the page should be _Mutable File-Systems (MFS)_, as should the sidebar nav item.
